besu/src/main/java/org/hyperledger/besu/cli/subcommands/storage/TrieLogHelper.java

```diff
@@ -0,0 +1,179 @@
+        totalToPrune, batchSize, numBatches);
+    out.println();
+
+    int noProgressCounter = 0;
+    int prevTotalNumberPruned = 0;
+    int totalNumberPruned = 0;
+    int numberPrunedInBatch;
+    int batchNumber = 1;
+    while (totalNumberPruned < totalToPrune) {
+      out.printf(
+          """
+          Pruning batch %d
+          -----------------
+          """, batchNumber++);
+      // do prune
+      numberPrunedInBatch = pruner.initialize();
+
+      out.printf("Number pruned in batch = %d \n", numberPrunedInBatch);
+      totalNumberPruned += numberPrunedInBatch;
+      out.printf(
+          """
+          Running total number pruned =
+          => %d of %d
+          """,
+          totalNumberPruned, totalToPrune);
+
+      if (totalNumberPruned == prevTotalNumberPruned) {
+        if (noProgressCounter++ == 5) {
+          out.println("No progress in 5 batches, exiting");
+          return;
+        }
+      }
+
+      prevTotalNumberPruned = totalNumberPruned;
+      out.println();
+    }
+    out.println("Trie log prune complete!");
+    out.println();
+
+    out.println("Counting trie logs after prune...");
+    TrieLogHelper.printCount(
+        out, TrieLogHelper.getCount(rootWorldStateStorage, Integer.MAX_VALUE, blockchain));
+  }
+
+  private static void validatePruneConfiguration(final DataStorageConfiguration config) {
+    checkArgument(
+        config.getUnstable().getBonsaiTrieLogRetentionThreshold()
+            >= MINIMUM_BONSAI_TRIE_LOG_RETENTION_THRESHOLD,
+        String.format(
+            "--Xbonsai-trie-log-retention-threshold minimum value is %d",
+            MINIMUM_BONSAI_TRIE_LOG_RETENTION_THRESHOLD));
+    checkArgument(
+        config.getUnstable().getBonsaiTrieLogPruningLimit() > 0,
+        String.format(
+            "--Xbonsai-trie-log-pruning-limit=%d must be greater than 0",
+            config.getUnstable().getBonsaiTrieLogPruningLimit()));
+    checkArgument(
+        config.getUnstable().getBonsaiTrieLogPruningLimit()
+            > config.getUnstable().getBonsaiTrieLogRetentionThreshold(),
+        String.format(
+            "--Xbonsai-trie-log-pruning-limit=%d must greater than --Xbonsai-trie-log-retention-threshold=%d",
+            config.getUnstable().getBonsaiTrieLogPruningLimit(),
+            config.getUnstable().getBonsaiTrieLogRetentionThreshold()));
+  }
+
+  static TrieLogCount getCount(
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final int limit,
+      final Blockchain blockchain) {
+    final AtomicInteger total = new AtomicInteger();
+    final AtomicInteger canonicalCount = new AtomicInteger();
+    final AtomicInteger forkCount = new AtomicInteger();
+    final AtomicInteger orphanCount = new AtomicInteger();
+    rootWorldStateStorage
+        .streamTrieLogKeys(limit)
+        .map(Bytes32::wrap)
+        .map(Hash::wrap)
+        .forEach(
```

Comment:
I am guessing that __most__ users of this subcommand are going to want to prune pretty close to head.  Meaning there will be far more trielogs deleted than retained.  This implementation is safe, but probably going to be pretty slow in most cases.  

IMO we could/should add an alternate implementation that checks that the number of retained trielogs is below a certain reasonable threshold.  We copy those trielogs to an alternate column family, truncate the trielog family, and move the trielogs back into the newly truncated CF.  That should make this operation markedly faster for what I suspect will be the typical use case (offline pruning a huge number of trielogs)

Reply:
to be clear, I would start from the blockchain storage and get the latest 'x' hashes that we want to retain rather than streaming all of the keys in the trielog CF